src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java

```diff
@@ -1017,6 +1026,61 @@ public String toString() {
     public static class ClassType extends Type implements DeclaredType, LoadableConstant,
                                                           javax.lang.model.type.ErrorType {
 
+        /**
+         * The 'flavor' of a ClassType indicates its reference/primitive projectionness
+         * viewed against the default nature of the associated class.
+         */
+        public enum Flavor {
```

Comment:
Perhaps you could add a method which "upgrades" a flavor to a new one, given new info.

e.g.

```
L_TypeOf_X.update(sym.isPrimitiveClass()) ->  L_TypeOf_Q/L_TypeOf_L (depending on isPrimitiveClass)
```

This would allow you to rule out (e.g. throw) bad updates. E.g. X flavors can be updated, all the other "leaf" flavors can only be updated to self. 

I think a method such as this would remove a lot of conditional logic in the clients.

Reply:
I'll take a look ...

Reply:
> Perhaps you could add a method which "upgrades" a flavor to a new one, given new info.
> 
> e.g.
> 
> ```
> L_TypeOf_X.update(sym.isPrimitiveClass()) ->  L_TypeOf_Q/L_TypeOf_L (depending on isPrimitiveClass)
> ```
> 
> This would allow you to rule out (e.g. throw) bad updates. E.g. X flavors can be updated, all the other "leaf" flavors can only be updated to self.
> 
> I think a method such as this would remove a lot of conditional logic in the clients.

Thanks, this turned out to be a nice suggestion and helped simplify the code at 3 or 4 places.